.github/flaky_test_detection.py

```diff
@@ -0,0 +1,114 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+import requests
+from github import Github
+from github import Auth
+
+
+GIT_ORG = "apache"
+GIT_REPO = "beam"
+GRAFANA_URL = "http://metrics.beam.apaache.org"
+#
+ALERT_NAME = "flaky_test"
+READ_ONLY = os.environ.get("READ_ONLY", "false")
+
+
+class Alert:
+    def __init__(self, workflow_id, workflow_url, workflow_name, workflow_filename):
+        self.workflow_id = workflow_id
+        self.workflow_url = workflow_url
+        self.workflow_name = workflow_name
+        self.workflow_file_name = workflow_filename
+
+
+def extract_workflow_id_from_issue_label(issues):
+    label_ids = []
+    for issue in issues:
+        for label in issue.get_labels():
+            match = re.search(r"workflow_id:\s*(\d+)", str(label.name))
+            if match:
+                label_id = match.group(1)
+                label_ids.append(label_id)
+
+    return label_ids
+
+
+def create_github_issue(repo, alert):
+    failing_runs_url = f"https://github.com/{GIT_ORG}/beam/actions/{alert.workflow_file_name}?query=is%3Afailure+branch%3Amaster"
+    title = f"The {alert.workflow_name} job is flaky"
+    body = f"The {alert.workflow_name } is constantly failing.\nPlease visit {failing_runs_url} to see the logs."
+    labels = ["flaky_test", f"workflow_id: {alert.workflow_id}", "bug", "P1"]
+    print("___")
+    print(f"Title: {title}")
+    print(f"Body: {body}")
+    print(f"Labels: {labels}")
+    print("___")
+
+    if READ_ONLY == "true":
+        print("READ_ONLY is true, not creating issue")
+    else:
+        repo.create_issue(title=title, body=body, labels=labels)
+
+
+def get_grafana_alerts():
+    url = f"{GRAFANA_URL}/api/alertmanager/grafana/api/v2/alerts?active=true&filter=alertname%3D{ALERT_NAME}"
+    response = requests.get(url)
+    if response.status_code != 200:
+        raise RuntimeError(
+            "Request to %s failed with status %d: %s"
+            % (url, response.status_code, response.text)
+        )
+    alerts = []
+    if response.text:
+        for alert in response.json():
+            alerts.append(
+                Alert(
+                    alert["labels"]["workflow_id"],
+                    alert["labels"]["workflow_url"],
+                    alert["labels"]["workflow_name"],
+                    alert["labels"]["workflow_filename"],
+                )
+            )
+    return alerts
+
+
+def main():
+    if "GITHUB_TOKEN" not in os.environ:
+        print("Please set the GITHUB_TOKEN environment variable.")
+        return
+
+    token = os.environ["GITHUB_TOKEN"]
+    auth = Auth.Token(token)
+    g = Github(auth=auth)
+    repo = g.get_repo(f"{GIT_ORG}/{GIT_REPO}")
+
+    alerts = get_grafana_alerts()
+    open_issues = repo.get_issues(state="open", labels=["flaky_test"])
```

Comment:
Hm, one problem I just thought of with this approach is that we will reopen issues after they are closed. For example, lets say:
1) 15 runs in a row have been failures so we filed an issue (since we've exceeded our 50% threshold for the last 30 runs).
2) Another 10 consecutive runs are failures. Now 25/30 runs are failures.
3) I fix the issue and get 3 consecutive green runs. Now 25/30 are still failures
4) I close the issue that I fixed.

Today, the issue would be reopened.

There are a few ways to address this problem, a few that come to mind are:
1) Only reopen the issue if we get at least 1 red status in the last 3 runs (and add instructions to only close the issue after 3 consecutive runs). Probabilistically, this should still catch most problems, but is prone to temporal issues (e.g. a workflow that only fails during periods of high traffic when the workflow runs at night).
2) Filter on issues that have been closed for less than a week. So basically, wait a week before reopening issues. We could even comment in the existing issue tagging the closer (saying something like "issue was closed, but alert is still firing, will reopen after XX days).

I'd lean towards (2) since its relatively simple (can be controlled all from this file). Basically, what I'd prefer is:
1) If an open issue exists for an alert, ignore it entirely.
2) If a closed issue (which has been closed for <7 days) exists for an alert, comment in the issue, tagging the issue closer.
3) If a closed issue (which has been closed for 7-14 days) exists for an alert, reopen the issue, tagging the issue closer.
4) If no issue exists or a closed issue has been closed for >14 days, open a new issue.

Thoughts?

Reply:
Great catch! +1 for (2), but we can improve this approach:
1. Open an issue If the number of failed runs exceeds the threshold
2. If an open issue exists for an alert, ignore it entirely
3. If someone closed an issue and it exists for an alert, check the `closed_at` issue field (https://docs.github.com/en/rest/issues/events?apiVersion=2022-11-28#get-an-issue-event) and filter out all runs before this date (we store the run [timestamp](https://github.com/apache/beam/pull/30327/files#diff-a99a5e5ac1dc4bf7f4d42360354deec95414d660122bd96214d190b92fee2340R386)). This will help us consider only "fresh" runs after the issue is closed.

Regarding re-opening vs opening a new issue - I would prefer to open a new issue every time we see deviations just to observe how the entire solution works for the Beam repo and improve it later if necessary.

Reply:
Sounds good to me.

> I would prefer to open a new issue every time we see deviations just to observe how the entire solution works for the Beam repo and improve it later if necessary.

I think that is fine for v1; I think reopening is also helpful for the case where the assignee didn't actually successfully fix the bug, but I'm fine adding this in later if needed (shouldn't be too hard).

Reply:
This would be a bit tricky but still i think it is possible. Regarding open/reopen ill try to see how reopening and adding a comment works but it seems like a better idea. Regarding the issue at hand its probably solvable with some math including time of first (first of 30 runs) vs time closed, we need to check it out. 
Regarding  Andreys comment we do currently:
1) open an issue if the number of failed runs exceeds the threshold
2) We do not open an issue if the issue with the same `workflow_id` label exists (only open at this time but the same mechanism can be used to reopen)

Reply:
Sounds good.

I'm going to merge this PR now since I think its taking good steps in the right direction, and we can address these issues:

1. not reopening alerts after the issue is fixed
2. reopening recently closed issues rather than opening new ones

in a follow up pr or two.